src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -2240,10 +2281,19 @@ WorkerExtendedOpNode(MultiExtendedOp *originalOpNode,
 									  &queryHavingQual, &queryTargetList,
 									  &queryGroupClause);
 
-	ProcessWindowFunctionsForWorkerQuery(originalWindowClause, originalTargetEntryList,
-										 &queryWindowClause, &queryTargetList);
+	if (!extendedOpNodeProperties->hasNonPushableWindowFunction)
+	{
+		ProcessWindowFunctionsForWorkerQuery(originalWindowClause,
+											 originalTargetEntryList,
+											 &queryWindowClause, &queryTargetList);
+	}
+	else
+	{
+		ProcessWindowFunctionPullUpForWorkerQuery(originalOpNode, &queryTargetList);
+	}
 
-	if (!extendedOpNodeProperties->pullUpIntermediateRows)
+	if (!extendedOpNodeProperties->hasNonPushableWindowFunction &&
```

Comment:
This could be more precise. Calling `ProcessDistinctClauseForWorkerQuery` is only an issue if one of the columns being distincted on involves a window function